drivers/mbox/mbox_ti_omap.c

```diff
@@ -0,0 +1,194 @@
+	uint32_t received_data;
+	void *mapped_base;
+};
+
+struct omap_mailbox_config {
+	uint32_t usr_id;
+};
+
+struct omap_mailbox_irq_regs {
+	uint32_t	status_raw;
+	uint32_t	status_clear;
+	uint32_t	enable_set;
+	uint32_t	enable_clear;
+};
+
+struct omap_mailbox_regs {
+	uint32_t	revision;
+	uint32_t	__pad0[3];
+	uint32_t	sysconfig;
+	uint32_t	__pad1[11];
+	uint32_t	message[OMAP_MAILBOX_NUM_MSGS];
+	uint32_t	fifo_status[OMAP_MAILBOX_NUM_MSGS];
+	uint32_t	msg_status[OMAP_MAILBOX_NUM_MSGS];
+	struct omap_mailbox_irq_regs irq_regs[OMAP_MAILBOX_NUM_USERS];
+};
+
+static void omap_mailbox_isr(const struct device *dev)
+{
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+	const struct omap_mailbox_config *cfg = DEV_CFG(dev);
+
+	volatile struct omap_mailbox_regs *regs = DEV_BASE(data);
+	volatile struct omap_mailbox_irq_regs *irq_regs = &regs->irq_regs[cfg->usr_id];
+	uint32_t flags = irq_regs->status_clear;
+
+	for (int i_channel = 0; i_channel < MAILBOX_MAX_CHANNELS; i_channel++) {
+		if (!data->channel_enable[i_channel])
+			continue;
+
+		if ((flags & MAILBOX_IRQ_NEWMSG(i_channel))) {
+			data->received_data = regs->message[i_channel];
+			struct mbox_msg msg = {(const void *)&data->received_data,
+					       MAILBOX_MBOX_SIZE};
+
+			if (data->cb[i_channel]) {
+				data->cb[i_channel](dev, i_channel, data->user_data[i_channel],
+						    &msg);
+				irq_regs->status_clear = MAILBOX_IRQ_NEWMSG(i_channel);
+			}
+
+		}
+	}
+}
+
+static int omap_mailbox_send(const struct device *dev, uint32_t channel, const struct mbox_msg *msg)
+{
+	uint32_t __aligned(4) data32 = 0;
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+	volatile struct omap_mailbox_regs *regs = DEV_BASE(data);
+
+	if (channel >= MAILBOX_MAX_CHANNELS)
+		return -EINVAL;
+
+	if (!msg) {
+		if (!regs->fifo_status[channel])
+			regs->message[channel] = 0;
+
+		return 0;
+	}
+
+	if (msg->size > MAILBOX_MBOX_SIZE)
+		return -EMSGSIZE;
+
+	memcpy(&data32, msg->data, msg->size);
+
+	if (!regs->fifo_status[channel])
+		regs->message[channel] = data32;
+
+	return 0;
+}
+
+static int omap_mailbox_register_callback(const struct device *dev, uint32_t channel,
+					  mbox_callback_t cb, void *user_data)
+{
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+
+	if (channel >= MAILBOX_MAX_CHANNELS)
+		return -EINVAL;
+
+	data->cb[channel] = cb;
+	data->user_data[channel] = user_data;
+
+	return 0;
+}
+
+static int omap_mailbox_mtu_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return MAILBOX_MBOX_SIZE;
+}
+
+static uint32_t omap_mailbox_max_channels_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return MAILBOX_MAX_CHANNELS;
+}
+
+static int omap_mailbox_set_enabled(const struct device *dev, uint32_t channel, bool enable)
+{
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+	const struct omap_mailbox_config *cfg = DEV_CFG(dev);
+	volatile struct omap_mailbox_regs *regs = DEV_BASE(data);
+	uint32_t irqstatus;
+
+	if (channel >= MAILBOX_MAX_CHANNELS)
+		return -EINVAL;
+
```

Comment:
This API is [supposed to return](https://docs.zephyrproject.org/latest/hardware/peripherals/mbox.html#c.mbox_set_enabled) `-EALREADY` if the channel is already enabled.

Since you're anyway tracking enabled channel via `channel_enable`:

```suggestion

	if (data->enabled_channels[channel]) {
		return -EALREADY;
	}

```